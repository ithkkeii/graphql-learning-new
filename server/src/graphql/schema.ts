const { gql } = require("apollo-server-express");

export const typeDefs = gql`
  # interface MutationResponse {
  #   code: String!
  #   success: Boolean!
  #   message: String!
  # }

  type AuthPayload {
    token: String!
    user: User!
  }

  type TestCase {
    text: String!
    testString: String!
  }

  type Challenge {
    id: ID!
    title: String!
    content: String!
    level: String!
    points: Int!
    contest: Contest!
    testCases: [TestCase!]!
    testInputs: [String!]!
    challengeSeed: String!
    passedUser: [User!]!
  }

  type User {
    id: ID!
    username: String!
    password: String
    firstname: String!
    lastname: String!
    solvedChallenges: [Challenge!]!
    likedChallenges: [Challenge!]!
    totalPoints: Int!
  }

  type Contest {
    id: ID!
    name: String!
    challenges: [Challenge!]!
  }

  type Query {
    getChallenge(id: ID!): Challenge
    getChallenges: [Challenge!]!
    getUser(id: ID!): User
    getUsers: [User!]!
    getContest(id: ID!): Contest!
    getContests: [Contest!]!
  }

  input UserInput {
    username: String!
    password: String!
    firstname: String!
    lastname: String!
  }

  input TestCaseInput {
    text: String!
    testString: String!
  }

  input ChallengeInput {
    title: String!
    content: String!
    level: String!
    points: Int!
    contestId: ID!
    testCases: [TestCaseInput!]!
    testInputs: [String!]!
    challengeSeed: String!
  }

  input ContestInput {
    name: String!
  }

  type Mutation {
    login(username: String!, password: String!): String!
    register(user: UserInput!): AuthPayload!
    editUser(userId: ID!, user: UserInput!): User!
    addChallenge(challenge: ChallengeInput!): Challenge!
    editChallenge(challengeId: ID!, challenge: ChallengeInput!): Challenge!
    addOrRemoveSolvedChallenges(userId: ID!, challengeId: ID!): [ID!]!
    addOrRemoveLikedChallenges(userId: ID!, challengeId: ID!): [ID!]!
    addContest(contest: ContestInput!): Contest!
    editContest(contestId: ID!, contest: ContestInput!): Contest!
    submitAnswer(challengeId: ID!, answer: String!): [String]!
  }
`;
