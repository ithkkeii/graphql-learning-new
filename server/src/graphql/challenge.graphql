type Challenge {
  id: ID!
  title: String!
  content: String!
  level: String!
  points: Int!
  category: Category!
  passedUser: [User!]!
}

type User {
  id: ID!
  username: String!
  password: String
  firstname: String!
  lastname: String!
  solvedChallenges: [Challenge!]!
  likedChallenges: [Challenge!]!
  totalPoints: Int!
}

type Category {
  id: ID!
  name: String!
  challenges: [Challenge!]!
}
