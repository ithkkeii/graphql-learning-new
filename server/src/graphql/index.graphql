# import Challenge from './challenge.graphql'
# import User from './challenge.graphql'
# import Category from './challenge.graphql'

type Query {
  getChallenge(id: ID!): Challenge
  getChallenges: [Challenge!]!
  getUser(id: ID!): User
  getUsers: [User!]!
  getCategory: Category!
  getCategories: [Category!]!
}

input UserInput {
  username: String!
  password: String!
  firstname: String!
  lastname: String!
}

input ChallengeInput {
  title: String!
  content: String!
  level: String!
  categoryId: ID!
}

input CategoryInput {
  name: String!
}

type Mutation {
  addUser(user: UserInput!): User!
  editUser(userId: ID!, user: UserInput!): User!
  addChallenge(challenge: ChallengeInput!): Challenge!
  editChallenge(challengeId: ID!, challenge: ChallengeInput!): Challenge!
  addOrRemoveSolvedChallenges(userId: ID!, challengeId: ID!): [ID!]!
  addOrRemoveLikedChallenges(userId: ID!, challengeId: ID!): [ID!]!
  addCategory(category: CategoryInput!): Category!
  editCategory(categoryId: ID!, category: CategoryInput!): Category!
}
