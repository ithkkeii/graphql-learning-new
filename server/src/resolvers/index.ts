import userResolvers from "./userResolvers";
import challengeResolvers from "./challengeResolvers";
import contestResolvers from "./contestResolvers";

const resolvers = {
  Query: {
    ...challengeResolvers.Query,
    ...userResolvers.Query,
    ...contestResolvers.Query,
  },
  User: userResolvers.User,
  Challenge: challengeResolvers.Challenge,
  Contest: contestResolvers.Contest,
  Mutation: {
    ...userResolvers.Mutation,
    ...challengeResolvers.Mutation,
    ...contestResolvers.Mutation,
  },
};

export default resolvers;
