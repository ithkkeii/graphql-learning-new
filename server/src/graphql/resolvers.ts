import mongoose, { mongo } from "mongoose";
import { Challenge, validateChallenge } from "../models/challenge";
import { User, validateUser } from "../models/user";
import { Contest, validateContest } from "../models/contest";
import { sign } from "jsonwebtoken";
import { authenticateUser } from "../utils/auth";
import { assert } from "chai";
import { writeFile, createWriteStream, write } from "fs";
import { join } from "path";
import { code } from "../code";
const bcrypt = require("bcrypt");

const resolvers = {
  Query: {
    getChallenge: async (parent: any, args: any, context: any) => {
      return await Challenge.findById(args.id);
    },
    getChallenges: async (parent: any, args: any, context: any) => {
      return await Challenge.find({});
    },
    //! Need admin authorization
    getUser: async (parent: any, args: any, context: any) => {
      return await User.findById(args.id);
    },
    //! Need admin authorization
    getUsers: async (parent: any, args: any, context: any) => {
      const user = await authenticateUser(context);
      return await User.find({});
    },
    getContest: async (parent: any, args: any, context: any) => {
      return await Contest.findById(args.id);
    },
    getContests: async (parent: any, args: any, context: any) => {
      return await Contest.find({});
    },
  },
  User: {
    solvedChallenges: async (
      parent: any,
      args: any,
      context: any,
      info: any
    ) => {
      const getChallenge = async (challengeId: String) => {
        return await Challenge.findById(challengeId);
      };

      const getChallenges = async () => {
        return Promise.all(
          parent.solvedChallenges.map((challengeId: any) =>
            getChallenge(challengeId)
          )
        );
      };

      return await getChallenges();
    },
    likedChallenges: async (
      parent: any,
      args: any,
      context: any,
      info: any
    ) => {
      const getChallenge = async (challengeId: String) => {
        return await Challenge.findById(challengeId);
      };

      const getChallenges = async () => {
        return Promise.all(
          parent.likedChallenges.map((challengeId: any) =>
            getChallenge(challengeId)
          )
        );
      };

      return await getChallenges();
    },
  },
  Challenge: {
    passedUser: async (parent: any, args: any, context: any, info: any) => {
      return await User.find({ solvedChallenges: parent.id }, "-password");
    },
    contest: async (parent: any, args: any, context: any, info: any) => {
      return await Contest.findById(parent.contest);
    },
  },
  Contest: {
    challenges: async (parent: any, args: any, context: any, info: any) => {
      return await Challenge.find({ contest: parent.contest });
    },
  },
  Mutation: {
    login: async (parent: any, args: any, context: any, info: any) => {
      const { username, password } = args;

      const { error } = validateUser({
        username,
        password,
        firstname: "validFirstName",
        lastname: "validLastName",
      });
      if (error) throw new Error(error.details[0].message);

      const user = await User.findOne({ username });
      if (!user) throw new Error(`Invalid username or password`);

      const isValidPassword = await bcrypt.compare(password, user.password);
      if (!isValidPassword) throw new Error(`Invalid username or password`);

      //! Import config for privatekey later & generateAuthToken need in user schema
      const token: string = sign(
        {
          _id: user._id,
          firstname: user.firstname,
          lastname: user.lastname,
        },
        "helloworld"
      );
      return token;
    },
    register: async (parent: any, args: any, context: any, info: any) => {
      const { username, password, firstname, lastname } = args.user;

      const { error } = validateUser(args.user);
      if (error) throw new Error(error.details[0].message);

      // If username already exist
      let user = await User.findOne({ username });
      if (user) throw new Error(`Username already taken`);

      user = new User({
        username,
        password,
        firstname,
        lastname,
      });

      user.password = await bcrypt.hash(user.password, 10);

      //? Do i need include password into token ?
      //! Import config for privatekey later & generateAuthToken need in user schema
      const token: string = sign(
        {
          _id: user._id,
          firstname: user.firstname,
          lastname: user.lastname,
        },
        "helloworld"
      );
      user = await user.save();
      delete user.password;

      return { token, user };
    },
    // ? How to create flexible InputUser graphql
    editUser: async (parent: any, args: any, context: any, info: any) => {
      const { userId } = args;
      const { username = "", password = "", firstname, lastname } = args.user;

      const { error } = validateUser(args.user);
      if (error) throw new Error(error.details[0].message);

      if (!mongoose.Types.ObjectId.isValid(userId))
        throw new Error(`Invalid user's id`);

      let user = await User.findById(userId);
      if (!user) throw new Error(`Invalid user's id`);

      user.firstname = firstname;
      user.lastname = lastname;
      user = await user.save();
      delete user.password;

      return user.save();
    },
    addChallenge: async (parent: any, args: any, context: any, info: any) => {
      const {
        title,
        content,
        level,
        points,
        contestId,
        testCases,
        testInputs,
        challengeSeed,
      } = args.challenge;

      const { error } = validateChallenge(args.challenge);
      if (error) throw new Error(error.details[0].message);

      const contest = await Contest.findById(contestId);
      if (!contest) throw new Error(`Invalid contest's id`);

      // If title exist
      let challenge = await Challenge.findOne({ title });
      if (challenge) throw new Error(`Title is already taken`);

      challenge = new Challenge({
        title,
        content,
        level,
        points,
        contest: contestId,
        testCases,
        testInputs,
        challengeSeed,
      });
      challenge = await challenge.save();

      return challenge;
    },
    editChallenge: async (parent: any, args: any, context: any, info: any) => {
      const { challengeId } = args;
      const {
        title,
        content,
        level,
        points,
        contestId,
        testCases,
        testInputs,
        challengeSeed,
      } = args.challenge;

      const { error } = validateChallenge(args.challenge);
      if (error) throw new Error(error.details[0].message);

      if (!mongoose.Types.ObjectId.isValid(challengeId))
        throw new Error(`Invalid challenge's id`);

      let challenge = await Challenge.findById(challengeId);
      if (!challenge) throw new Error(`Invalid challenge's id`);

      const uniqueTitleChallenge = await Challenge.findOne({
        title,
        _id: { $nin: [challengeId] },
      });
      if (uniqueTitleChallenge)
        throw new Error(`Challenge' title is already exist`);

      const contest = await Contest.findById(contestId);
      if (!contest) throw new Error(`Invalid contest's id`);

      challenge.title = title;
      challenge.content = content;
      challenge.level = level;
      challenge.points = points;
      challenge.contest = contestId;
      challenge.testCases = testCases;
      challenge.testInputs = testInputs;
      challenge.challengeSeed = challengeSeed;

      challenge = await challenge.save();
      return challenge;
    },
    // ! Dont need to remove solvedChallenges
    addOrRemoveSolvedChallenges: async (
      parent: any,
      args: any,
      context: any
    ) => {
      const { userId, challengeId } = args;

      if (!mongoose.Types.ObjectId.isValid(challengeId))
        throw new Error(`Invalid challenge's id`);

      if (!mongoose.Types.ObjectId.isValid(userId))
        throw new Error(`Invalid user's id`);

      let user = await User.findById(userId);
      if (!user) throw new Error(`Invalid user's id`);

      let challenge = await Challenge.findById(challengeId);
      if (!challenge) throw new Error(`Invalid challenge's id`);

      const index: number = user.solvedChallenges.indexOf(challengeId);
      if (index === -1) {
        user.solvedChallenges.push(challengeId);
      } else {
        user.solvedChallenges.splice(index, 1);
      }
      user = await user.save();
      delete user.password;
      return user.solvedChallenges;
    },
    addOrRemoveLikedChallenges: async (
      parent: any,
      args: any,
      context: any
    ) => {
      const { userId, challengeId } = args;

      if (!mongoose.Types.ObjectId.isValid(challengeId))
        throw new Error(`Invalid challenge's id`);

      if (!mongoose.Types.ObjectId.isValid(userId))
        throw new Error(`Invalid user's id`);

      let user = await User.findById(userId);
      if (!user) throw new Error(`Invalid user's id`);

      let challenge = await Challenge.findById(challengeId);
      if (!challenge) throw new Error(`Invalid challenge's id`);

      const index: number = user.likedChallenges.indexOf(challengeId);
      if (index === -1) {
        user.likedChallenges.push(challengeId);
      } else {
        user.likedChallenges.splice(index, 1);
      }
      user = await user.save();
      delete user.password;
      return user.likedChallenges;
    },
    addContest: async (parent: any, args: any, context: any, info: any) => {
      const { name } = args.contest;

      const { error } = validateContest(args.contest);
      if (error) throw new Error(error.details[0].message);

      // Check name unique
      let contest = await Contest.findOne({ name });
      if (contest) throw new Error(`Name of contest is already taken`);

      contest = new Contest({ name });
      return await contest.save();
    },
    editContest: async (parent: any, args: any, context: any, info: any) => {
      const { name } = args.contest;

      const { error } = validateContest(args.contest);
      if (error) throw new Error(error.details[0].message);

      if (!mongoose.Types.ObjectId.isValid(args.contestId))
        throw new Error(`Invalid contest's id`);

      let contest = await Contest.findById(args.contestId);
      if (!contest) throw new Error(`Invalid contest's id`);

      // Check name unique
      let uniqueContest = await Contest.findOne({
        name,
        _id: { $nin: [args.contestId] },
      });
      if (uniqueContest) throw new Error(`Name of contest is already taken`);

      contest.name = name;
      return await contest.save();
    },
    submitAnswer: async (parent: any, args: any, context: any, info: any) => {
      const { challengeId, answer } = args;

      const sum: Function = new Function(answer)();
      const checkAnonymous: Function = new Function(
        "assert",
        "sum",
        "return assert(sum(1,2) === 3)"
      );

      // await writeFile("test.ts", code, () => {});
      const writer = createWriteStream("test.ts", { encoding: "utf8" });
      writer.write(code);
      // if (anonymousFunc()(1, 2) === 3) return [true];

      // console.log(checkAnonymous(assert, sum).toString());
      console.log(sum.toString());
      return [true];
    },
  },
};

export default resolvers;
