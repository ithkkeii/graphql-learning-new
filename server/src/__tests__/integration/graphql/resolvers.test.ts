import { gql, ApolloServer } from "apollo-server-express";
const { createTestClient } = require("apollo-server-testing");
import { typeDefs } from "../../../graphql/schema";
import resolvers from "../../../graphql/resolvers";
import mongooseServer from "../../../db";

// Connect to mongoDB
mongooseServer();

const GET_USERS = gql`
  query GetUsers {
    getUsers {
      id
      username
      password
      firstname
      lastname
      totalPoints
      solvedChallenges {
        id
        title
        content
        level
        points
        contest {
          id
          name
          challenges {
            id
            title
          }
        }
        passedUser {
          id
          username
          password
          firstname
          lastname
          totalPoints
        }
      }
      likedChallenges {
        id
        title
        content
        level
        points
        contest {
          id
          name
          challenges {
            id
            title
          }
        }
        passedUser {
          id
          username
          password
          firstname
          lastname
          totalPoints
        }
      }
    }
  }
`;

const GET_USER = gql`
  query GetUser($id: ID!) {
    getUser(id: $id) {
      id
      username
      password
      firstname
      lastname
      totalPoints
      solvedChallenges {
        id
        title
        content
        level
        points
        contest {
          id
          name
          challenges {
            id
            title
          }
        }
        passedUser {
          id
          username
          password
          firstname
          lastname
          totalPoints
        }
      }
      likedChallenges {
        id
        title
        content
        level
        points
        contest {
          id
          name
          challenges {
            id
            title
          }
        }
        passedUser {
          id
          username
          password
          firstname
          lastname
          totalPoints
        }
      }
    }
  }
`;

const GET_CHALLENGE = gql`
  query GetChallenge($id: ID!) {
    getChallenge(id: $id) {
      id
      title
      content
      level
      points
      contest {
        id
        name
        challenges {
          id
        }
      }
      testCases {
        text
        testString
      }
      testInputs
      challengeSeed
      passedUser {
        id
        username
        password
        firstname
        lastname
        solvedChallenges {
          id
        }
        likedChallenges {
          id
        }
        totalPoints
      }
    }
  }
`;

const GET_CHALLENGES = gql`
  query GetChallenges {
    getChallenges {
      id
      title
      content
      level
      points
      contest {
        id
        name
        challenges {
          id
        }
      }
      testCases {
        text
        testString
      }
      testInputs
      challengeSeed
      passedUser {
        id
        username
        password
        firstname
        lastname
        solvedChallenges {
          id
        }
        likedChallenges {
          id
        }
        totalPoints
      }
    }
  }
`;

const GET_CONTESTS = gql`
  query GetContests {
    getContests {
      id
      name
      challenges {
        id
        title
        content
        level
        points
        contest {
          id
          name
          challenges {
            id
          }
        }
        testCases {
          text
          testString
        }
        testInputs
        challengeSeed
        passedUser {
          id
          username
          password
          firstname
          lastname
          solvedChallenges {
            id
          }
          likedChallenges {
            id
          }
          totalPoints
        }
      }
    }
  }
`;

const GET_CONTEST = gql`
  query GetContest($id: ID!) {
    getContest(id: $id) {
      id
      name
      challenges {
        id
        title
        content
        level
        points
        contest {
          id
          name
          challenges {
            id
          }
        }
        testCases {
          text
          testString
        }
        testInputs
        challengeSeed
        passedUser {
          id
          username
          password
          firstname
          lastname
          solvedChallenges {
            id
          }
          likedChallenges {
            id
          }
          totalPoints
        }
      }
    }
  }
`;

const apolloServer = new ApolloServer({ typeDefs, resolvers });
const { query, mutate } = createTestClient(apolloServer);

describe("Queries", () => {
  describe("User", () => {
    it("should return all users", async () => {
      const res = await query({ query: GET_USERS });
      expect(res).toMatchSnapshot();
    });

    it("should return single user with provided ID", async () => {
      const res = await query({
        query: GET_USER,
        variables: { id: "5eaa773df4dcc728b76b00b7" },
      });
      expect(res).toMatchSnapshot();
    });
  });

  describe("Challenge", () => {
    it("should return all challenges", async () => {
      const res = await query({ query: GET_CHALLENGES });
      expect(res).toMatchSnapshot();
    });

    it("should return single challenge with provided ID", async () => {
      const res = await query({
        query: GET_CHALLENGE,
        variables: { id: "5eaa7d1effd50d33c34637a8" },
      });
      expect(res).toMatchSnapshot();
    });
  });

  describe("Contest", () => {
    it("should return all contests", async () => {
      const res = await query({
        query: GET_CONTESTS,
      });
      expect(res).toMatchSnapshot();
    });

    it("should return single contest with provided ID", async () => {
      const res = await query({
        query: GET_CONTEST,
        variables: { id: "5eaa779bf4dcc728b76b00bb" },
      });
      expect(res).toMatchSnapshot();
    });
  });
});
