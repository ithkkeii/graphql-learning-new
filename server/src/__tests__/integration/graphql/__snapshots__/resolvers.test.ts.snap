// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries Challenge should return all challenges 1`] = `
Object {
  "data": Object {
    "getChallenges": Array [
      Object {
        "challengeSeed": "'use strict'",
        "content": "John is playing with his friends in the park",
        "contest": Object {
          "challenges": Array [
            Object {
              "id": "5eaa7d1effd50d33c34637a8",
            },
            Object {
              "id": "5eaa7f442db1ce3470fa16cc",
            },
          ],
          "id": "5eaa779bf4dcc728b76b00bb",
          "name": "Sherlock Homes",
        },
        "id": "5eaa7d1effd50d33c34637a8",
        "level": "easy",
        "passedUser": Array [
          Object {
            "firstname": "kenny",
            "id": "5eaa773df4dcc728b76b00b7",
            "lastname": "uchiha",
            "likedChallenges": Array [
              Object {
                "id": "5eaa7d1effd50d33c34637a8",
              },
            ],
            "solvedChallenges": Array [
              Object {
                "id": "5eaa7d1effd50d33c34637a8",
              },
            ],
            "totalPoints": 0,
            "username": "kenny",
          },
        ],
        "points": 100,
        "testCases": Array [
          Object {
            "testString": "assert.strictEqual(result,10)",
            "text": "Must be 10",
          },
        ],
        "testInputs": Array [
          "3\\\\n1 2 10",
        ],
        "title": "Find max",
      },
      Object {
        "challengeSeed": "'use strict'",
        "content": "John is playing with his friends in the park",
        "contest": Object {
          "challenges": Array [
            Object {
              "id": "5eaa7d1effd50d33c34637a8",
            },
            Object {
              "id": "5eaa7f442db1ce3470fa16cc",
            },
          ],
          "id": "5eaa779bf4dcc728b76b00bb",
          "name": "Sherlock Homes",
        },
        "id": "5eaa7f442db1ce3470fa16cc",
        "level": "easy",
        "passedUser": Array [],
        "points": 100,
        "testCases": Array [
          Object {
            "testString": "assert.strictEqual(result,1)",
            "text": "Must be 1",
          },
        ],
        "testInputs": Array [
          "3\\\\n1 2 10",
        ],
        "title": "Find min",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries Challenge should return single challenge with provided ID 1`] = `
Object {
  "data": Object {
    "getChallenge": Object {
      "challengeSeed": "'use strict'",
      "content": "John is playing with his friends in the park",
      "contest": Object {
        "challenges": Array [
          Object {
            "id": "5eaa7d1effd50d33c34637a8",
          },
          Object {
            "id": "5eaa7f442db1ce3470fa16cc",
          },
        ],
        "id": "5eaa779bf4dcc728b76b00bb",
        "name": "Sherlock Homes",
      },
      "id": "5eaa7d1effd50d33c34637a8",
      "level": "easy",
      "passedUser": Array [
        Object {
          "firstname": "kenny",
          "id": "5eaa773df4dcc728b76b00b7",
          "lastname": "uchiha",
          "likedChallenges": Array [
            Object {
              "id": "5eaa7d1effd50d33c34637a8",
            },
          ],
          "solvedChallenges": Array [
            Object {
              "id": "5eaa7d1effd50d33c34637a8",
            },
          ],
          "totalPoints": 0,
          "username": "kenny",
        },
      ],
      "points": 100,
      "testCases": Array [
        Object {
          "testString": "assert.strictEqual(result,10)",
          "text": "Must be 10",
        },
      ],
      "testInputs": Array [
        "3\\\\n1 2 10",
      ],
      "title": "Find max",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries Contest should return all contests 1`] = `
Object {
  "data": Object {
    "getContests": Array [
      Object {
        "challenges": Array [],
        "id": "5eaa7772f4dcc728b76b00b8",
        "name": "Warm up!",
      },
      Object {
        "challenges": Array [],
        "id": "5eaa7780f4dcc728b76b00b9",
        "name": "Solve the function",
      },
      Object {
        "challenges": Array [],
        "id": "5eaa7787f4dcc728b76b00ba",
        "name": "The big map",
      },
      Object {
        "challenges": Array [
          Object {
            "challengeSeed": "'use strict'",
            "content": "John is playing with his friends in the park",
            "contest": Object {
              "challenges": Array [
                Object {
                  "id": "5eaa7d1effd50d33c34637a8",
                },
                Object {
                  "id": "5eaa7f442db1ce3470fa16cc",
                },
              ],
              "id": "5eaa779bf4dcc728b76b00bb",
              "name": "Sherlock Homes",
            },
            "id": "5eaa7d1effd50d33c34637a8",
            "level": "easy",
            "passedUser": Array [
              Object {
                "firstname": "kenny",
                "id": "5eaa773df4dcc728b76b00b7",
                "lastname": "uchiha",
                "likedChallenges": Array [
                  Object {
                    "id": "5eaa7d1effd50d33c34637a8",
                  },
                ],
                "solvedChallenges": Array [
                  Object {
                    "id": "5eaa7d1effd50d33c34637a8",
                  },
                ],
                "totalPoints": 0,
                "username": "kenny",
              },
            ],
            "points": 100,
            "testCases": Array [
              Object {
                "testString": "assert.strictEqual(result,10)",
                "text": "Must be 10",
              },
            ],
            "testInputs": Array [
              "3\\\\n1 2 10",
            ],
            "title": "Find max",
          },
          Object {
            "challengeSeed": "'use strict'",
            "content": "John is playing with his friends in the park",
            "contest": Object {
              "challenges": Array [
                Object {
                  "id": "5eaa7d1effd50d33c34637a8",
                },
                Object {
                  "id": "5eaa7f442db1ce3470fa16cc",
                },
              ],
              "id": "5eaa779bf4dcc728b76b00bb",
              "name": "Sherlock Homes",
            },
            "id": "5eaa7f442db1ce3470fa16cc",
            "level": "easy",
            "passedUser": Array [],
            "points": 100,
            "testCases": Array [
              Object {
                "testString": "assert.strictEqual(result,1)",
                "text": "Must be 1",
              },
            ],
            "testInputs": Array [
              "3\\\\n1 2 10",
            ],
            "title": "Find min",
          },
        ],
        "id": "5eaa779bf4dcc728b76b00bb",
        "name": "Sherlock Homes",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries Contest should return single contest with provided ID 1`] = `
Object {
  "data": Object {
    "getContest": Object {
      "challenges": Array [
        Object {
          "challengeSeed": "'use strict'",
          "content": "John is playing with his friends in the park",
          "contest": Object {
            "challenges": Array [
              Object {
                "id": "5eaa7d1effd50d33c34637a8",
              },
              Object {
                "id": "5eaa7f442db1ce3470fa16cc",
              },
            ],
            "id": "5eaa779bf4dcc728b76b00bb",
            "name": "Sherlock Homes",
          },
          "id": "5eaa7d1effd50d33c34637a8",
          "level": "easy",
          "passedUser": Array [
            Object {
              "firstname": "kenny",
              "id": "5eaa773df4dcc728b76b00b7",
              "lastname": "uchiha",
              "likedChallenges": Array [
                Object {
                  "id": "5eaa7d1effd50d33c34637a8",
                },
              ],
              "solvedChallenges": Array [
                Object {
                  "id": "5eaa7d1effd50d33c34637a8",
                },
              ],
              "totalPoints": 0,
              "username": "kenny",
            },
          ],
          "points": 100,
          "testCases": Array [
            Object {
              "testString": "assert.strictEqual(result,10)",
              "text": "Must be 10",
            },
          ],
          "testInputs": Array [
            "3\\\\n1 2 10",
          ],
          "title": "Find max",
        },
        Object {
          "challengeSeed": "'use strict'",
          "content": "John is playing with his friends in the park",
          "contest": Object {
            "challenges": Array [
              Object {
                "id": "5eaa7d1effd50d33c34637a8",
              },
              Object {
                "id": "5eaa7f442db1ce3470fa16cc",
              },
            ],
            "id": "5eaa779bf4dcc728b76b00bb",
            "name": "Sherlock Homes",
          },
          "id": "5eaa7f442db1ce3470fa16cc",
          "level": "easy",
          "passedUser": Array [],
          "points": 100,
          "testCases": Array [
            Object {
              "testString": "assert.strictEqual(result,1)",
              "text": "Must be 1",
            },
          ],
          "testInputs": Array [
            "3\\\\n1 2 10",
          ],
          "title": "Find min",
        },
      ],
      "id": "5eaa779bf4dcc728b76b00bb",
      "name": "Sherlock Homes",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries User should return all users 1`] = `
Object {
  "data": Object {
    "getUsers": Array [
      Object {
        "firstname": "kenny",
        "id": "5eaa773df4dcc728b76b00b7",
        "lastname": "uchiha",
        "likedChallenges": Array [
          Object {
            "content": "John is playing with his friends in the park",
            "contest": Object {
              "challenges": Array [
                Object {
                  "id": "5eaa7d1effd50d33c34637a8",
                  "title": "Find max",
                },
                Object {
                  "id": "5eaa7f442db1ce3470fa16cc",
                  "title": "Find min",
                },
              ],
              "id": "5eaa779bf4dcc728b76b00bb",
              "name": "Sherlock Homes",
            },
            "id": "5eaa7d1effd50d33c34637a8",
            "level": "easy",
            "passedUser": Array [
              Object {
                "firstname": "kenny",
                "id": "5eaa773df4dcc728b76b00b7",
                "lastname": "uchiha",
                "totalPoints": 0,
                "username": "kenny",
              },
            ],
            "points": 100,
            "title": "Find max",
          },
        ],
        "solvedChallenges": Array [
          Object {
            "content": "John is playing with his friends in the park",
            "contest": Object {
              "challenges": Array [
                Object {
                  "id": "5eaa7d1effd50d33c34637a8",
                  "title": "Find max",
                },
                Object {
                  "id": "5eaa7f442db1ce3470fa16cc",
                  "title": "Find min",
                },
              ],
              "id": "5eaa779bf4dcc728b76b00bb",
              "name": "Sherlock Homes",
            },
            "id": "5eaa7d1effd50d33c34637a8",
            "level": "easy",
            "passedUser": Array [
              Object {
                "firstname": "kenny",
                "id": "5eaa773df4dcc728b76b00b7",
                "lastname": "uchiha",
                "totalPoints": 0,
                "username": "kenny",
              },
            ],
            "points": 100,
            "title": "Find max",
          },
        ],
        "totalPoints": 0,
        "username": "kenny",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries User should return single user with provided ID 1`] = `
Object {
  "data": Object {
    "getUser": Object {
      "firstname": "kenny",
      "id": "5eaa773df4dcc728b76b00b7",
      "lastname": "uchiha",
      "likedChallenges": Array [
        Object {
          "content": "John is playing with his friends in the park",
          "contest": Object {
            "challenges": Array [
              Object {
                "id": "5eaa7d1effd50d33c34637a8",
                "title": "Find max",
              },
              Object {
                "id": "5eaa7f442db1ce3470fa16cc",
                "title": "Find min",
              },
            ],
            "id": "5eaa779bf4dcc728b76b00bb",
            "name": "Sherlock Homes",
          },
          "id": "5eaa7d1effd50d33c34637a8",
          "level": "easy",
          "passedUser": Array [
            Object {
              "firstname": "kenny",
              "id": "5eaa773df4dcc728b76b00b7",
              "lastname": "uchiha",
              "totalPoints": 0,
              "username": "kenny",
            },
          ],
          "points": 100,
          "title": "Find max",
        },
      ],
      "solvedChallenges": Array [
        Object {
          "content": "John is playing with his friends in the park",
          "contest": Object {
            "challenges": Array [
              Object {
                "id": "5eaa7d1effd50d33c34637a8",
                "title": "Find max",
              },
              Object {
                "id": "5eaa7f442db1ce3470fa16cc",
                "title": "Find min",
              },
            ],
            "id": "5eaa779bf4dcc728b76b00bb",
            "name": "Sherlock Homes",
          },
          "id": "5eaa7d1effd50d33c34637a8",
          "level": "easy",
          "passedUser": Array [
            Object {
              "firstname": "kenny",
              "id": "5eaa773df4dcc728b76b00b7",
              "lastname": "uchiha",
              "totalPoints": 0,
              "username": "kenny",
            },
          ],
          "points": 100,
          "title": "Find max",
        },
      ],
      "totalPoints": 0,
      "username": "kenny",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
